integration-tests/src/tests/client/process_blocks.rs

```diff
@@ -74,479 +73,501 @@ use std::collections::{HashSet, VecDeque};
+            if let NetworkRequests::Approval { .. } = msg.as_network_requests_ref() {
+                let mut first_header_announce = first_header_announce.write();
+                if *first_header_announce {
+                    *first_header_announce = false;
                 } else {
-                    EpochId(last_block.header.next_epoch_id)
-                },
-                None,
-                vec![],
-                Ratio::from_integer(0),
-                0,
-                100,
-                None,
-                &signer,
-                last_block.header.next_bp_hash,
-                block_merkle_tree.root(),
-                Clock::real(),
-                None,
-                None,
-                None,
-            );
-            actor_handles.client_actor.send(
-                BlockResponse {
-                    block: block.into(),
-                    peer_id: PeerInfo::random().id,
-                    was_requested: false,
+                    done_clone.cancel();
+                    actor_system.stop();
                 }
-                .span_wrap(),
-            );
-            future::ready(())
-        });
-        actix::spawn(actor);
-        near_network::test_utils::wait_or_panic(5000);
-    });
+            }
+            PeerManagerMessageResponse::NetworkResponses(NetworkResponses::NoResponse)
+        }),
+    );
+    let res = actor_handles.view_client_actor.send_async(GetBlockWithMerkleTree::latest()).await;
+    let (last_block, block_merkle_tree) = res.unwrap().unwrap();
+    let mut block_merkle_tree = PartialMerkleTree::clone(&block_merkle_tree);
+    block_merkle_tree.insert(last_block.header.hash);
+    let signer = create_test_signer("test1");
+    let next_block_ordinal = last_block.header.block_ordinal.unwrap() + 1;
+    let block = Block::produce(
+        PROTOCOL_VERSION,
+        &last_block.header.clone().into(),
+        last_block.header.height + 1,
+        next_block_ordinal,
+        last_block.chunks.iter().cloned().map(Into::into).collect(),
+        vec![vec![]; last_block.chunks.len()],
+        EpochId::default(),
+        if last_block.header.prev_hash == CryptoHash::default() {
+            EpochId(last_block.header.hash)
+        } else {
+            EpochId(last_block.header.next_epoch_id)
+        },
+        None,
+        vec![],
+        Ratio::from_integer(0),
+        0,
+        100,
+        None,
+        &signer,
+        last_block.header.next_bp_hash,
+        block_merkle_tree.root(),
+        Clock::real(),
+        None,
+        None,
+        None,
+    );
+    actor_handles.client_actor.send(
+        BlockResponse { block: block.into(), peer_id: PeerInfo::random().id, was_requested: false }
+            .span_wrap(),
+    );
+    tokio::select! {
+        _ = done.cancelled() => {},
+        _ = tokio::time::sleep(tokio::time::Duration::from_millis(5000)) => panic!("Timeout exceeded while waiting for block approval."),
+    }
```

Comment:
dumb question: Why is there no case for something like done.completed? Is the typical flow to have it cancelled?